### PR TITLE
Add support for training apis to support custom ops

### PR DESCRIPTION
**Pull request by baijumeswani**
So far, the training API did not include support to register custom op libraries.

One of the models we are working with requires the use of `onnxruntime-extensions`. `onnxruntime-extensions` is implemented as a custom op and needs to be registered with `ONNX Runtime` before an ONNX model containing the custom op nodes can be executed.

This pull request adds the necessary tooling to register the custom op library for both, artifact generation, and graph execution.

In addition, this pull request also adds support for executing models with string tensor inputs.

### CODE REVIEW

**Review by pengwa — COMMENTED** [review]
I am curious how custom ops are needed in on device training scenarios? the product team want to write their own operator?

File: orttraining/orttraining/python/training/onnxblock/model_accessor.py [hunk 1]
```diff
@@ -115,3 +117,32 @@ def empty_base(opset_version: Optional[int] = None):
...
+        custom_op_library_path: The path to the custom op library.
+
+    Returns:
+        ModelAccessor: The model accessor that contains the modified model.
+    """
+    global _GLOBAL_CUSTOM_OP_LIBRARY  # pylint: disable=global-statement  # noqa: PLW0603
+    if _GLOBAL_CUSTOM_OP_LIBRARY is not None:
+        raise RuntimeError("CustomOp library already set. Cannot set multiple custom op libraries.")
+
+    _GLOBAL_CUSTOM_OP_LIBRARY = copy.copy(custom_op_library_path)  # noqa: PLW0603
```
Comment (pengwa) [inline orttraining/orttraining/python/training/onnxblock/model_accessor.py:144]: do you think we should have a check here in case of wasting many work before failing on path not valid?

File: orttraining/orttraining/python/training/onnxblock/_training_graph_utils.py [hunk 2]
```diff
@@ -106,10 +116,14 @@ def build_gradient_graph(
     eval_model = copy.deepcopy(model)
     _disable_training_mode(eval_model)
 
-    optimized_model = onnx.load_from_string(get_optimized_model(model.SerializeToString(), requires_grad))
+    options = SessionOptions()
+    if custom_op_library is not None:
```
Comment (pengwa) [inline orttraining/orttraining/python/training/onnxblock/_training_graph_utils.py:120]: curious why we need set this before build gradient graph.

**Comment by baijumeswani** [comment]
> I am curious how custom ops are needed in on device training scenarios? the product team want to write their own operator?

One of our partner teams has a need for the op [`StringToHashFast` defined in `onnxruntime-extensions`](https://github.com/microsoft/onnxruntime-extensions/blob/main/docs/custom_ops.md#stringhashfast). In order to execute an op defined in `onnxruntime-extensions` library, the library must be registered with `ONNX Runtime` as a custom library.

This pull request allows for us to register any user provided custom op library (such as `onnxruntime-extensions`.

File: orttraining/orttraining/python/training/onnxblock/_training_graph_utils.py (hunk 2 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/_training_graph_utils.py:120]: The graph optimization and gradient graph building, both require us to load the ONNX model into ONNX Runtime's `Model` class. However, loading an ONNX model with custom op definitions will fail since ONNX Runtime cannot understand which op the node corresponds to (without having access to the schema).
This is why we need to pass in the custom op library to ONNX Runtime while optimizing the graph and building the gradient graph.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 3]
```diff
@@ -167,25 +168,36 @@ struct TrainingConfigurationResult {
...
+  std::unique_ptr<GradientGraphConfiguration> gradient_graph_config_;
+  std::shared_ptr<CustomRegistry> custom_registry_;
+  IOnnxRuntimeOpSchemaRegistryList local_registries_;
+  PyGradientGraphBuilder(std::unique_ptr<GradientGraphBuilder> builder,
+                         std::shared_ptr<Model> model, std::unique_ptr<logging::Logger> logger,
+                         std::unique_ptr<GradientGraphConfiguration> gradient_graph_config,
+                         std::shared_ptr<CustomRegistry> custom_registry,
+                         IOnnxRuntimeOpSchemaRegistryList local_registries)
+      : builder_(std::move(builder)),
+        model_(std::move(model)),
```
Comment (pengwa) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:196]:
```suggestion
        model_(std::move(model)),
```

since model is passed as param from caller, is std::move appropriate for the shared_ptr?

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 4]
```diff
@@ -167,25 +168,36 @@ struct TrainingConfigurationResult {
...
-  PyGradientGraphBuilder(std::unique_ptr<GradientGraphBuilder> builder_, std::shared_ptr<Model> model_, std::unique_ptr<logging::Logger> logger_, std::unique_ptr<GradientGraphConfiguration> gradient_graph_config_)
-      : builder(std::move(builder_)), model(std::move(model_)), logger(std::move(logger_)), gradient_graph_config(std::move(gradient_graph_config_)) {}
+  std::unique_ptr<GradientGraphBuilder> builder_;
+  std::shared_ptr<Model> model_;
+  std::unique_ptr<logging::Logger> logger_;
+  std::unique_ptr<GradientGraphConfiguration> gradient_graph_config_;
+  std::shared_ptr<CustomRegistry> custom_registry_;
+  IOnnxRuntimeOpSchemaRegistryList local_registries_;
+  PyGradientGraphBuilder(std::unique_ptr<GradientGraphBuilder> builder,
+                         std::shared_ptr<Model> model, std::unique_ptr<logging::Logger> logger,
```
Comment (pengwa) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:191]: nit:

```suggestion
                         std::shared_ptr<Model> model, 
                         std::unique_ptr<logging::Logger> logger,
```

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 5]
```diff
@@ -812,42 +824,58 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
-                              *logger_ptr);
-
-                          return std::make_unique<PyGradientGraphBuilder>(std::move(builder), std::move(model), std::move(logger_ptr), std::move(gradient_graph_config_ptr));
-                        }))
+  gradient_graph_builder
+      .def(py::init([](const py::bytes& serialized_model,
+                       const std::unordered_set<std::string>& y_node_arg_names,
+                       const std::unordered_set<std::string>& x_node_arg_names,
+                       const std::string loss_node_arg_name,
+                       PySessionOptions* options) {
```
Comment (pengwa) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:832]: I assume this means ORTModule also has the opportunity to support ort-extension later using this one.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 6]
```diff
@@ -812,42 +824,58 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
+             IOnnxRuntimeOpSchemaRegistryList local_registries;
+             if (options && !options->custom_op_domains_.empty()) {
+               // Register all custom op domains that will be needed for the session
+               ORT_THROW_IF_ERROR(onnxruntime::CreateCustomRegistry(options->custom_op_domains_, custom_registry));
+               local_registries.push_back(custom_registry->GetOpschemaRegistry());
+             }
+
+             std::shared_ptr<Model> model;
+             auto logger_ptr = std::make_unique<logging::Logger>(logging::LoggingManager::DefaultLogger());
+             logger_ptr->SetSeverity(logging::Severity::kINFO);
```
Comment (pengwa) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:843]: if we pass session options, shall we parse log_severity_level from it and use it instead of hard coding to info?

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 7]
```diff
@@ -167,25 +168,36 @@ struct TrainingConfigurationResult {
...
-  std::unique_ptr<GradientGraphConfiguration> gradient_graph_config;
-  PyGradientGraphBuilder(std::unique_ptr<GradientGraphBuilder> builder_, std::shared_ptr<Model> model_, std::unique_ptr<logging::Logger> logger_, std::unique_ptr<GradientGraphConfiguration> gradient_graph_config_)
-      : builder(std::move(builder_)), model(std::move(model_)), logger(std::move(logger_)), gradient_graph_config(std::move(gradient_graph_config_)) {}
+  std::unique_ptr<GradientGraphBuilder> builder_;
+  std::shared_ptr<Model> model_;
+  std::unique_ptr<logging::Logger> logger_;
+  std::unique_ptr<GradientGraphConfiguration> gradient_graph_config_;
+  std::shared_ptr<CustomRegistry> custom_registry_;
+  IOnnxRuntimeOpSchemaRegistryList local_registries_;
+  PyGradientGraphBuilder(std::unique_ptr<GradientGraphBuilder> builder,
```
Comment (pengwa) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:190]: `PyGradientGraphBuilder ` did not looks to be gradient graph builder only things, it is more like a `PyGradientGraphBuildContext` which has the builder, and also other configs .

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 8]
```diff
@@ -885,23 +913,60 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
                        std::optional<std::string> eval_model_uri,
-                       OrtDevice device) {
-        onnxruntime::SessionOptions session_option;
+                       OrtDevice device, PySessionOptions* session_options) {
         std::vector<std::shared_ptr<IExecutionProvider>> provider = GetExecutionProvidersForTrainingApis(device);
-
         auto env = GetTrainingEnv().GetORTEnv();
-        return std::make_unique<onnxruntime::training::api::Module>(
-            model_uri, state, session_option, *env, provider, eval_model_uri);
+        auto model = std::make_unique<onnxruntime::training::api::Module>(
```
Comment (pengwa) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:919]: nit: why not just return without assignment to local varl as before?

**Review by pengwa — COMMENTED** [review]
Here are some comments.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 5 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:834]: You're right. This can be extended to be used with ORTModule.
But we can even go further by allowing users to register custom gradient kernels in a custom op library if needed.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 8 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/orttraining_pybind_state.cc]: Thanks. I was playing around locally and forgot to revert back later.

File: orttraining/orttraining/training_api/training_session.h [hunk 9]
```diff
@@ -29,7 +29,8 @@ class TrainingSession {
                   const SessionOptions& session_options,
                   const std::vector<std::shared_ptr<IExecutionProvider>>& providers,
                   CheckpointState* state,
-                  const ModelIdentifiers& model_identifiers);
+                  const ModelIdentifiers& model_identifiers,
+                  gsl::span<OrtCustomOpDomain* const> custom_op_domains);
```
Comment (pengwa) [inline orttraining/orttraining/training_api/training_session.h:33]: shall we give a default value too for training_session as we did for module and optimizer? 

```suggestion
                  gsl::span<OrtCustomOpDomain* const> custom_op_domains = gsl::span<OrtCustomOpDomain* const>());
```

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 10]
```diff
@@ -885,23 +919,58 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
-              const std::vector<OrtValue>& user_inputs, std::vector<OrtValue>& user_outputs) -> void {
-             ORT_THROW_IF_ERROR(model->EvalStep(user_inputs, user_outputs));
+              const std::vector<py::object>& user_inputs, std::vector<OrtValue>& user_outputs) -> void {
+             std::vector<OrtValue> feeds;
+             const auto model_inputs_with_error = model->GetEvalModelInputs();
+             ORT_THROW_IF_ERROR(model_inputs_with_error.first);
+             ORT_ENFORCE(model_inputs_with_error.second, "Eval model graph inputs are not defined.");
+             for (size_t idx = 0; idx < user_inputs.size(); ++idx) {
+               auto& feed = user_inputs[idx];
+               const auto feed_name = model->GetEvalModelInputName(idx);
```
Comment (pengwa) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:961]: same here.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 11]
```diff
@@ -885,23 +919,58 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
+             for (size_t idx = 0; idx < user_inputs.size(); ++idx) {
+               auto& feed = user_inputs[idx];
+               const auto feed_name = model->GetTrainingModelInputName(idx);
+               // No need to process 'None's sent in by the user
+               // to feed Optional inputs in the graph.
+               // We just won't include anything in the feed and ORT
+               // will handle such implicit 'None's internally.
+               if (!feed.is(py::none())) {
+                 OrtValue ort_value;
+                 CreateGenericMLValue(model_inputs_with_error.second, GetAllocator(), feed_name, feed, &ort_value);
```
Comment (pengwa) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:945]: hmmm, then it will be CPU only input.

File: orttraining/orttraining/test/python/orttraining_test_python_bindings.py [hunk 12]
```diff
@@ -368,3 +371,118 @@ def test_get_input_output_names():
...
+
+        training_model_file_path = pathlib.Path(temp_dir) / "training_model.onnx"
+        eval_model_file_path = pathlib.Path(temp_dir) / "eval_model.onnx"
+        checkpoint_file_path = pathlib.Path(temp_dir) / "checkpoint"
+
+        # Create Checkpoint State.
+        state = CheckpointState.load_checkpoint(checkpoint_file_path)
+
+        # Create a Module.
+        # The custom op library is built either for cuda or cpu (but not for both).
```
Comment (askhade) [inline orttraining/orttraining/test/python/orttraining_test_python_bindings.py:424]: thanks for these comments :)

File: orttraining/orttraining/python/training/api/module.py [hunk 13]
```diff
@@ -70,17 +74,7 @@ def __call__(self, *user_inputs) -> tuple[np.ndarray] | np.ndarray:
         Returns:
             The outputs of the model.
         """
-        is_np_input = False
-        forward_inputs = OrtValueVector()
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/python/training/api/module.py:74]: I'll add comments to the code.
The reason I changed this is because there is no way to create an OrtValue from a numpy ndarray of strings. The op `StringToHashBucket` takes in a tensor of strings as input. But the earliest implementation did not permit us to execute a model where the input is a tensor of strings.

With this change, the input can be either an numpy ndarray, or a OrtValue (or even a native python list). See this function for details:
https://github.com/microsoft/onnxruntime/blob/2de5807703cb1553cc299ecd203cd531ab62e77b/onnxruntime/python/onnxruntime_pybind_mlvalue.cc#L824-L896

File: orttraining/orttraining/python/training/api/module.py [hunk 14]
```diff
@@ -70,17 +74,7 @@ def __call__(self, *user_inputs) -> tuple[np.ndarray] | np.ndarray:
...
-        forward_inputs.reserve(len(user_inputs))
-        for tensor in user_inputs:
-            if isinstance(tensor, np.ndarray):
-                is_np_input = True
-                forward_inputs.push_back(OrtValue.ortvalue_from_numpy(tensor)._ortvalue)
-            elif isinstance(tensor, OrtValue):
-                forward_inputs.push_back(tensor._ortvalue)
-            else:
-                raise ValueError(f"Expected input of type: numpy array or OrtValue, actual: {type(tensor)}")
+        forward_inputs = [user_input for user_input in user_inputs]
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/python/training/api/module.py:77]: Yes, to convert tuple to list. I can do `list(user_inputs)` if that is clearer.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 11 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:945]: While the input is CPU only, the graph can be executed on cuda if the execution provider is cuda.

This is the primary difference between run with io binding and run in the onnxruntime inference realm.

But you're right. We should also allow users to execute the graph with inputs on gpu. Let me try to support both scenarios in this PR.

File: orttraining/orttraining/python/training/onnxblock/blocks.py [hunk 15]
```diff
@@ -427,3 +428,45 @@ def build(self, cast_input_name: str):
         self.base.graph.node.append(cast_node)
 
         return cast_output_name
+
+
+class Gemm(Block):
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/blocks.py:433]: I think I can add those on as needed basis.

Since the use case for this `Gemm` is similar to `torch.nn.Linear` (maybe I should call it Linear instead of Gemm). Since the users are introducing a new node in the graph, they only specify the `in_features` and the `out_features`. As such, they don't get to chose the shape of the weight. So `transB` is not very meaningful.

As for `transA`, it is mostly needed for optimizing the graph. I can add this as needed later on as needed.

File: orttraining/orttraining/python/training/api/module.py [hunk 16]
```diff
@@ -59,42 +62,57 @@ def __init__(
             state._state,
             os.fspath(eval_model_uri) if eval_model_uri is not None else None,
             self._device,
+            self._session_options,
         )
         self._state = state
 
-    def __call__(self, *user_inputs) -> tuple[np.ndarray] | np.ndarray:
+    def __call__(self, *user_inputs) -> tuple[np.ndarray] | np.ndarray | tuple[OrtValue] | OrtValue:
```
Comment (pengwa) [inline orttraining/orttraining/python/training/api/module.py:69]: the tuple usage looks not correct. for variance length, it is tuple[str, ...]); for two elements, it's tuple[str, str]

```suggestion
    def __call__(self, *user_inputs) -> tuple[np.ndarray, ...] | np.ndarray | tuple[OrtValue, ...] | OrtValue:
```

**Review by pengwa — APPROVED** [review]
Looks good, there is one minor comments.

**Review by pengwa — APPROVED** [review]
